File: src/finance_journal/repositories/impostazioni.py
import sqlite3
from datetime import date


class ImpostazioniRepository:
    def __init__(self, conn: sqlite3.Connection) -> None:
        self._conn = conn

    def get(self, chiave: str, default: str | None = None) -> str | None:
        row = self._conn.execute(
            "SELECT valore FROM impostazioni WHERE chiave = ?", (chiave,)
        ).fetchone()
        return row["valore"] if row is not None else default

    def set(self, chiave: str, valore: str) -> None:
        self._conn.execute(
            "INSERT INTO impostazioni (chiave, valore) VALUES (?, ?)"
            " ON CONFLICT(chiave) DO UPDATE SET valore = excluded.valore",
            (chiave, valore),
        )
        self._conn.commit()

    def get_saldo_iniziale(self) -> tuple[float, date | None]:
        importo = 0.0
        importo_str = self.get("saldo_iniziale_importo")
        if importo_str:
            try:
                importo = float(importo_str)
            except ValueError:
                pass
        saldo_data: date | None = None
        data_str = self.get("saldo_iniziale_data")
        if data_str:
            try:
                saldo_data = date.fromisoformat(data_str)
            except ValueError:
                pass
        return importo, saldo_data

File: electron/src/ipc/impostazioni.ts
import type Database from 'better-sqlite3';

export const CHIAVI = {
  SALDO_INIZIALE_IMPORTO: 'saldo_iniziale_importo',
  SALDO_INIZIALE_DATA: 'saldo_iniziale_data',
} as const;

export function getSaldoIniziale(db: Database.Database, anno: number): number {
  try {
    const importoStr = getImpostazione(db, CHIAVI.SALDO_INIZIALE_IMPORTO);
    const dataStr = getImpostazione(db, CHIAVI.SALDO_INIZIALE_DATA);
    if (!importoStr || !dataStr || dataStr.length < 4) return 0;
    const annoSaldo = parseInt(dataStr.slice(0, 4), 10);
    if (isNaN(annoSaldo) || annoSaldo > anno) return 0;
    const importo = parseFloat(importoStr);
    return isNaN(importo) ? 0 : importo;
  } catch (err) {
    throw new Error(`Failed to get saldo iniziale: ${String(err)}`);
  }
}

export function getImpostazione(db: Database.Database, chiave: string): string | null {
  try {
    const row = db
      .prepare('SELECT valore FROM impostazioni WHERE chiave = ?')
      .get(chiave) as { valore: string } | undefined;
    return row?.valore ?? null;
  } catch (err) {
    throw new Error(`Failed to get impostazione "${chiave}": ${String(err)}`);
  }
}

export function setImpostazione(db: Database.Database, chiave: string, valore: string): void {
  try {
    db.prepare(`
      INSERT INTO impostazioni (chiave, valore, updated_at)
      VALUES (?, ?, datetime('now'))
      ON CONFLICT(chiave) DO UPDATE SET
        valore     = excluded.valore,
        updated_at = excluded.updated_at
    `).run(chiave, valore);
  } catch (err) {
    throw new Error(`Failed to set impostazione "${chiave}": ${String(err)}`);
  }
}
